### ISSUE TITLE
Full sync crashes with "Script aborted: 'names'"

### ISSUE THREAD

**Issue by Turochamp** [antonplagemann/GoogleMonicaSync#15]
Looks like a great tool that would make Monica so much better for me. But cannot get it to work... yet.

Following the 'Get started' and doing the last sync step I get this output:

> Your choice (unordered):
> - initial sync
> 
> Fetching Google contacts...
> Finished fetching Google contacts
> Fetching all Monica contacts (page 11 of 11)
> Finished fetching Monica contacts
> 
> Building sync database...
> Processing Google contact 442 of 1012
> Script aborted: 'names'

Any suggestions how to diagnose this further?

**Comment by Turochamp** [comment]
Got a debugger running and captured the exception:

  File "/home/michael/git/GoogleMonicaSync/GMSync.py", line 92, in main
    sync.startSync('initial')
  File "/home/michael/git/GoogleMonicaSync/SyncHelper.py", line 44, in startSync
    self.__initialSync()
  File "/home/michael/git/GoogleMonicaSync/SyncHelper.py", line 81, in __initialSync
    self.__buildSyncDatabase()
  File "/home/michael/git/GoogleMonicaSync/SyncHelper.py", line 552, in __buildSyncDatabase
    monicaId = self.__simpleMonicaIdSearch(googleContact)
  File "/home/michael/git/GoogleMonicaSync/SyncHelper.py", line 1066, in __simpleMonicaIdSearch
    gContactGivenName = googleContact['names'][0].get("givenName", False)

During handling of the above exception, another exception occurred:

  File "/home/michael/git/GoogleMonicaSync/GMSync.py", line 122, in main
    sys.exit(1)
  File "/home/michael/git/GoogleMonicaSync/GMSync.py", line 126, in <module>
    main()

**Comment by antonplagemann** [comment]
Hi @Turochamp,
thanks for reporting this issue and all the details! 🏅 
Unfortunately I'm currently in vacation. I will take a deeper look when I'm back (propably sunday 🙂).
It seems like you have a contact without a name (given, middle or surname). I'm not sure how to handle this case, because for Monica contacts a name is mandatory. 
I made the (wrong) assumption that every Google contact has at least a name.
Because of this I now need to change a few things (not a quick 'one-line' fix).
Please be patient until next week, I will post any update here 🙂

**Comment by Turochamp** [comment]
Sounds great @antonplagemann! :-)
Let me know if there is anything else we can do to verify the diagnosis. Knowing is always better then guessing.

**Comment by antonplagemann** [comment]
Hi @Turochamp,
as I assumed, the error was caused by a missing name.
I implemented a fix for that in v3 but you need to update your config file to get it working.
As you can see in the readme, you have to add a `NAME_IF_UNNAMED = 'Unnamed contact'` line in the config.

How the fix works:
- If there is a contact without name during initial sync (`-i`) you will be asked to chose one of these options:
  - Input a custom name for this very contact
  - Input a default name for all contacts without a name
  - Use `NAME_IF_UNNAMED` constant from the config as custom name for this very contact
  - Use `NAME_IF_UNNAMED` constant from the config as default name for all contacts without a name
- If there is a contact without name during full or delta sync (`-f` or `-d`):
  - If it is a new contact the `NAME_IF_UNNAMED` constant from the config will be used as first name
  - If it is an existing (already synced) contact the name will NOT be synced, meaning the old name will be retained
- If you add a name to a contact that did not have a name before it will be synced again and overwrites the name at Monica

To sum up:
1. Update your repo by doing `git pull`
2. Update your config by adding `NAME_IF_UNNAMED = 'your default name here'`
3. Run `GMSync.py -i` and follow the prompts

I hope this fixes the error.
Thanks again for reporting it, if there are any questions left or you have additional suggestions I'll be happy to hear that 😃

**Comment by Turochamp** [comment]
Hi @antonplagemann,
This is awesome! Result of the fix is that `Sync.log` now includes a list of my 16 Google contact that don't have names. These I don't want in my Monica, it's not contacts that I care enough about to give a name so obviously they don't belong in my Monica database. To exclude these I started adding a label to these Google contacts, then I though "Wait a minute! This is unnecessary toil and the sync would be much better if this exclusion was automated".

Before continuing on that thought. Actually there is a bug in this fix as it fails when trying to print the unnamed contact. The list was only created after I commented out that line of code:

    def __handleEmptyGoogleNames(self, googleContact: dict) -> str:
        '''Ask the user for a new given name.'''
        # Ask user if not done before
        if not self.skipUnnamedPrompt:
            # print(self.__getGoogleContactAsString(googleContact))
            print(f"The contact above has no name, but Monica needs at least a first name.")
            print(f"\tWhat would you like to do?")
            
I'm happy to provide you with information how Google handles these contacts without name so the sync can print them the same way. 

So what do you want to do now? We can either 1) Add a automated exclusion of Google contacts that are not valid (nor useful) contacts in Monica or 2) Fix the bug in the current handle unnamed feature.

Personally I will never use the unnamed feature so I think 1) would be best. Though I agree it is unpleasant to remove code, I think it's better long term to have the best possible user experience (less toil) and code maintainability (less code).

**Comment by Turochamp** [comment]
Also @antonplagemann,
To help with testing I've created scripts to deploy a new Monica testing instance with a copy of my database. So I am now ready with a test environment including all my 1000 contacts.

Just let me know how I can help. Having this working will be very valuable for me.

**Comment by antonplagemann** [comment]
Hi @Turochamp,
thanks for all your thoughts and sorry for the delay!
> So what do you want to do now? We can either 1) Add a automated exclusion of Google contacts that are not valid (nor useful) contacts in Monica or 2) Fix the bug in the current handle unnamed feature.
How about both? 😄
I agree to your arguments that it would be a better user experience to skip these contacts.
But I think it will be useful to print a warning in the console and the log file, including the affected contact.
> I'm happy to provide you with information how Google handles these contacts without name so the sync can print them the same way.
That would be great! In this case I can rewrite the `__getGoogleContactAsString()` method to print them this way.
> Actually there is a bug in this fix as it fails when trying to print the unnamed contact.
It would be great if you have an error message or the log content of that bug for me (maybe its not neccessary anymore with the solution above, but I'm curious because i didn't find something :-) )
> To help with testing I've created scripts to deploy a new Monica testing instance with a copy of my database. So I am now ready with a test environment including all my 1000 contacts.
Thats cool! And more than I have... 😄

**Comment by Turochamp** [comment]
Hi @antonplagemann 
Both it is then. 👍 

I'll come back with further answers shortly.

**Comment by Turochamp** [comment]
Hi @antonplagemann 

Here is the program output and exception stack. The error is for my 7th unnamed contact, as I've excluded six unnamed contacts. 

Based upon these seven unnamed contacts there are two different ways Google prints name:
- Has company name: Google uses company name
- Only has email: Google uses email

```
Building sync database...
Processing Google contact 491 of 1012
Script aborted: 'contactGroups/myContacts'
```

```
  File "/home/michael/git/GoogleMonicaSync/GMSync.py", line 78, in main
    sync.startSync('initial')
  File "/home/michael/git/GoogleMonicaSync/SyncHelper.py", line 48, in startSync
    self.__initialSync()
  File "/home/michael/git/GoogleMonicaSync/SyncHelper.py", line 85, in __initialSync
    self.__buildSyncDatabase()
  File "/home/michael/git/GoogleMonicaSync/SyncHelper.py", line 556, in __buildSyncDatabase
    monicaId = self.__simpleMonicaIdSearch(googleContact)
  File "/home/michael/git/GoogleMonicaSync/SyncHelper.py", line 1099, in __simpleMonicaIdSearch
    gContactGivenName = self.__handleEmptyGoogleNames(googleContact)
  File "/home/michael/git/GoogleMonicaSync/SyncHelper.py", line 1153, in __handleEmptyGoogleNames
    print(self.__getGoogleContactAsString(googleContact))
  File "/home/michael/git/GoogleMonicaSync/SyncHelper.py", line 1222, in __getGoogleContactAsString
    name = self.google.getLabelName(value['contactGroupResourceName'])
  File "/home/michael/git/GoogleMonicaSync/GoogleHelper.py", line 70, in getLabelName
    return self.reverseLabelMapping[labelId]
```

**Comment by antonplagemann** [comment]
Hi @Turochamp 

thanks for the stack trace.
I implemented a fix and changed the name handling in `v3.1.1`.
I would be very happy, if you can confirm it working in your test environment 😎

**Comment by Turochamp** [comment]
Hi @antonplagemann, 
Tested and bug fixed. Thanks for the help!
